### ISSUE TITLE
Accessibility and assistive technologies

### ISSUE THREAD

**Issue by blindphoneman** [barotto/IBMulator#88]
![screenshot_0000](https://github.com/barotto/IBMulator/assets/45212938/455e3221-a3ea-41bc-a573-16cbd3c729c6)

Running Hitchhiker's Guide to the Galaxy from Lost Treasures of Infocom CD-Rom

Good job!

**Comment by barotto** [comment]
Thank you for sharing your experience!
I'm glad it's working well.

I was thinking about adding text-to-speech to the interface in the next version, so that IBMulator can tell to the user what the its GUI is doing, like what window is active or which element has focus. Do you think it would be helpful?

**Comment by blindphoneman** [comment]
Hi:

 

This is a fantastic idea.

 

Don’t know how far you want to go with it, but, here are a couple ideas:

 

The custom dialogs that come up are not accessible with my Windows screen readers, so, if they could be made to speak, that would be helpful.  I can use the “native” file dialogs, but, the custom dialogs allow more flexibility.

 

Here’s another idea:

 

I know the BIOS text screen is always available to the program.  Maybe that could be spoken.  One emulator developer wrote this into a TI 99/4A emulator called Classic 99.  His program installs similarly to Ibmulator, so you can run it and delete it after you’re through. He put a screen reading mode into the program that uses a dif algorithm to read the screen.  You can look at what he has done at https://www.harmlesslion.com/software/classic99.

 

A third possibility, and, my favorite idea, is to allow either a comport or parallel port to be configurable to speak anything sent through it.  You could then use a DOS screen reader to handle the data.  I currently have lpt1 directed to a file, and, in Windows 10, it lets mee use com4 as the file.  Windows 11 will not let me do this.  One day, I’ll have to move to Windows 11, if this PC dies.

 

Could a second comport just for speech be added when this is needed?  At least, in Windows, you could have data sent to one of the PS1’s comports spoken using the SAPI protocol, or maybe Windows 1core voices.  A couple screen readers can use “generic” synthesizers.  That way, you wouldn’t need to worry about control codes.  The main control code you’d need at a minimum would be to make the speech silence with ctrl-X.  Not permanent silence, just stops it from speaking what it already has in its buffers.

 

You may not do anything with these, but, here are links for 3 different screen readers.  Probably the most popular one in the day was either JAWS or Vocal-Eyes.  ASAP is my favorite one, and has the smallest footprint and uses only one command to start it.

 

Here are the links:

 

JAWS: https://nfbnet.org/files/blind/JDOSFREE.EXE

 

Vocal-Eyes: https://nfbnet.org/files/blind/VE.ZIP

 

Automatic Screen Access Program (ASAP): https://nfbnet.org/files/blind/ASAPDEMO.EXE

 

JAWS is Freeware, the other 2 are demos, but, cannot be purchased anymore.

 

This is more than you’ll need, but, maybe save it and see if any of these give you ideas to work with.

 

If you want to contact me, you can email me at joseph(dot)norton(at)gmail.com (just substitute the punctuation in parenthesis with the real thing).  I’ll be happy to test anything you come up with.

 

Thanks so much for thinking about this.

**Comment by blindphoneman** [comment]
Hi:

The link for Classic99 should be:

https://harmlesslion.com/software/classic99

It didn't like it with the www on it.

Sorry for the confusion.

**Comment by barotto** [comment]
Thank you for your help. I need to study the technologies and find the proper solutions. I'll definitely contact you for future testing and help on the matter!

**Comment by barotto** [comment]
Hello, a lot happened in the past year and I finally found the time to resume working on this project.

Also Accessibility turned out to be a bigger topic than I initially thought.

If you're still interested in testing, I would appeciate very much your help with the version of IBMulator I'm working on:

[IBMulator 0.19-alpha1](http://test.ibmulator.org/ibmulator-0.19-alpha1-windows.x86_64.zip)

Please consider that it's still a work in progress.  
I'm looking for feedback on the text-to-speech system in particular, but also on usability and documentation.

In the package there's a description of the system in the README.md file, under the Accessibility section.

This alpha version of IBMulator uses the SAPI 5 interface for the voice synthesizer and has an internal Braille 'n Speak emulator. There's no need for an external emulator.

I suggest you let the program create a new ibmulator.ini file as there's a new `[tts]` section.  
There are new TTS related key bindings as well, so an updated keymap.map file is also needed.

Or you can copy and paste this text in ibmulator.ini:

```
[tts]
;   gui_enabled: Enable Text-To-Speech for the GUI.
; guest_enabled: Enable Text-To-Speech for the guest OS.
;       gui_dev: The device to use for the GUI.
;                 Possible values: synth, file.
;                  synth: use a voice synthetizer.
;                   file: redirect to a file.
;         voice: The synthesizer's voice (see the README for more info).
;          file: The file name to use for the file mode.
;   file_format: The text format to use when file mode is used.
;                 Possible values: simple, ssml, msxml.
;                  simple: simple text.
;                    ssml: W3C's Speech Synthesis Markup Language.
;                   msxml: Microsoft's SAPI 5 synthesis markup.
;     guest_dev: The device to use for the guest OS.
;                 Possible values: auto, or one of the values of the 'gui_dev' setting.
;                  auto: use the same value used in the 'gui_dev' setting.
;      codepage: The code page of the guest OS.
;                 Common DOS code pages are 437 and 850. The default is 437.
gui_enabled=yes
guest_enabled=yes
gui_dev=synth
voice=default 
file=
file_format=
guest_dev=auto
codepage=437
```

And this text in keymap.map:

```
### TTS bindings

KMOD_CTRL + KMOD_SHIFT + SDLK_3 = FUNC_TTS_GUI_TOGGLE
KMOD_CTRL + KMOD_SHIFT + SDLK_4 = FUNC_TTS_GUEST_TOGGLE
KMOD_CTRL + KMOD_SHIFT + SDLK_X = FUNC_TTS_STOP
KMOD_CTRL + KMOD_SHIFT + SDLK_B = FUNC_TTS_DESCRIBE
KMOD_CTRL + KMOD_SHIFT + SDLK_T = FUNC_TTS_WINDOW_TITLE
KMOD_CTRL + KMOD_SHIFT + SDLK_1 = FUNC_TTS_READ_CHARS
KMOD_CTRL + KMOD_SHIFT + SDLK_2 = FUNC_TTS_READ_WORDS
KMOD_CTRL + KMOD_SHIFT + SDLK_A = FUNC_TTS_ADJ_VOLUME(1)
KMOD_CTRL + KMOD_SHIFT + SDLK_Z = FUNC_TTS_ADJ_VOLUME(-1)
KMOD_CTRL + KMOD_SHIFT + SDLK_D = FUNC_TTS_ADJ_RATE(1)
KMOD_CTRL + KMOD_SHIFT + SDLK_C = FUNC_TTS_ADJ_RATE(-1)
```
